src/services/AccessTokenRequestProcessor.ts

```diff
@@ -0,0 +1,127 @@
+
+export class AccessTokenRequestProcessor {
+	private static instance: AccessTokenRequestProcessor;
+
+	private readonly logger: Logger;
+
+	private readonly metrics: Metrics;
+
+	private readonly bavService: BavService;
+
+	private readonly kmsJwtAdapter: KmsJwtAdapter;
+
+	private readonly issuer: string;
+
+	constructor(logger: Logger, metrics: Metrics) {
+		this.logger = logger;
+		this.metrics = metrics;
+  		logger.debug("metrics is  " + JSON.stringify(this.metrics));
+  		this.metrics.addMetric("Called", MetricUnits.Count, 1);
+
+		const sessionTableName: string = checkEnvironmentVariable(EnvironmentVariables.SESSION_TABLE, this.logger);
+  		const signingKeyArn: string = checkEnvironmentVariable(EnvironmentVariables.KMS_KEY_ARN, this.logger);
+		this.issuer = checkEnvironmentVariable(EnvironmentVariables.ISSUER, this.logger);
+		
+		this.kmsJwtAdapter = new KmsJwtAdapter(signingKeyArn);
+		this.bavService = BavService.getInstance(sessionTableName, this.logger, createDynamoDbClient());
+	}
+
+	static getInstance(logger: Logger, metrics: Metrics): AccessTokenRequestProcessor {
+		if (!AccessTokenRequestProcessor.instance) {
+			AccessTokenRequestProcessor.instance = new AccessTokenRequestProcessor(logger, metrics);
+		}
+		return AccessTokenRequestProcessor.instance;
+	}
+
+	async processRequest(event: APIGatewayProxyEvent): Promise<Response> {		
+		try {
+			let requestPayload;
+			try {
+				requestPayload = isPayloadValid(event.body);
+			} catch (error) {
+				this.logger.error("Failed validating the Access token request body.", { messageCode: MessageCodes.FAILED_VALIDATING_ACCESS_TOKEN_REQUEST_BODY });
+				if (error instanceof AppError) {
+					return new Response(error.statusCode, error.message);
+				}
+				return new Response(HttpCodesEnum.UNAUTHORIZED, "An error has occurred while validating the Access token request payload.");
+			}
+			let session: ISessionItem | undefined;
+			try {
+				session = await this.bavService.getSessionByAuthorizationCode(requestPayload.code);
+				if (!session) {
+					this.logger.info(`No session found by authorization code: : ${requestPayload.code}`, { messageCode: MessageCodes.SESSION_NOT_FOUND });
+					return new Response(HttpCodesEnum.UNAUTHORIZED, `No session found by authorization code: ${requestPayload.code}`);
+				}
+				this.logger.appendKeys({ sessionId: session.sessionId });
+				this.logger.info({ message: "Found Session" });
+				this.logger.appendKeys({
+					govuk_signin_journey_id: session?.clientSessionId,
+				});
+			} catch (error) {
+				if (error instanceof AppError) {
+					return new Response(error.statusCode, error.message);
+				}
+
+				this.logger.error("Error while retrieving the session", {
+					messageCode: MessageCodes.SESSION_NOT_FOUND,
+					error,
+				});
+				return new Response(HttpCodesEnum.UNAUTHORIZED, "Error while retrieving the session");
+			}
+
+			//if (session.authSessionState === AuthSessionState.BAV_AUTH_CODE_ISSUED) {
```

Comment:
This check looks like it's in the `validateTokenRequestToRecord` function now, but I actually agree that it would be better here, and we leave that validate function to just validate one thing (redirectUri)